exotations/solvers/exotica_levenberg_marquardt_solver/src/LevenbergMarquardt.cpp

```diff
@@ -0,0 +1,138 @@
+#include <exotica/MotionSolver.h>
+#include <exotica/Problems/UnconstrainedEndPoseProblem.h>
+#include <exotica_lm_solver/LMSolverInitializer.h>
+
+namespace exotica
+{
+class LevenbergMarquardt : public MotionSolver, public Instantiable<LMSolverInitializer>
+{
+public:
+    virtual void Instantiate(LMSolverInitializer& init);
+
+    virtual void Solve(Eigen::MatrixXd& solution);
+
+    virtual void specifyProblem(PlanningProblem_ptr pointer);
+
+private:
+    LMSolverInitializer parameters_;
+
+    UnconstrainedEndPoseProblem_ptr prob_;  // Shared pointer to the planning problem.
+
+    double lambda = 0;  // damping factor
+
+    int iterations_ = -1;
+};
+
+REGISTER_MOTIONSOLVER_TYPE("LMSolver", exotica::LevenbergMarquardt)
+
+void LevenbergMarquardt::Instantiate(LMSolverInitializer& init) { parameters_ = init; }
+void LevenbergMarquardt::specifyProblem(PlanningProblem_ptr pointer)
+{
+    if (pointer->type() != "exotica::UnconstrainedEndPoseProblem")
+    {
+        throw_named("This LevenbergMarquardt can't solve problem of type '" << pointer->type() << "'!");
+    }
+
+    MotionSolver::specifyProblem(pointer);
+
+    // generic problem
+    problem_ = pointer;
+
+    // specific problem
+    prob_ = std::static_pointer_cast<UnconstrainedEndPoseProblem>(pointer);
+}
+
+void LevenbergMarquardt::Solve(Eigen::MatrixXd& solution)
+{
+    prob_->resetCostEvolution(getNumberOfMaxIterations() + 1);
+
+    Timer timer;
+
+    if (!prob_) throw_named("Solver has not been initialized!");
+
+    const Eigen::VectorXd q0 = prob_->applyStartState();
+
+    if (prob_->N != q0.rows()) throw_named("Wrong size q0 size=" << q0.rows() << ", required size=" << prob_->N);
+
+    solution.resize(1, prob_->N);
+
+    lambda = parameters_.Damping;  // initial damping
+
+    const Eigen::MatrixXd I = Eigen::MatrixXd::Identity(prob_->Cost.J.cols(), prob_->Cost.J.cols());
+    Eigen::MatrixXd J;
+
+    Eigen::VectorXd q = q0;
+    double error = std::numeric_limits<double>::infinity();
+    double error_prev = std::numeric_limits<double>::infinity();
+    Eigen::VectorXd yd;
+    Eigen::VectorXd qd;
+    for (size_t i = 0; i < getNumberOfMaxIterations(); iterations_ = ++i)
+    {
+        prob_->Update(q);
+
+        yd = prob_->Cost.S * prob_->Cost.ydiff;
```

Comment:
`Cost` may become a private member in the future.

Reply:
Will it have a getter `getCost()`?

Reply:
It will as needed - as evident by this solver - and `exotica_ik_solver`.